File: 04/demos/demo/src/main/java/com/pluralsight/dynamodb/dao/CommentDao.java
package com.pluralsight.dynamodb.dao;

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBQueryExpression;
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBScanExpression;
import com.amazonaws.services.dynamodbv2.model.AttributeValue;
import com.amazonaws.services.dynamodbv2.model.ComparisonOperator;
import com.amazonaws.services.dynamodbv2.model.Condition;
import com.amazonaws.services.dynamodbv2.transactions.Transaction;
import com.amazonaws.services.dynamodbv2.transactions.TransactionManager;
import com.pluralsight.dynamodb.domain.Comment;
import com.pluralsight.dynamodb.domain.Item;

import java.util.List;

public class CommentDao {

    private final DynamoDBMapper mapper;

    private final TransactionManager txManager;

    public CommentDao(AmazonDynamoDB dynamoDB) {
        this.mapper = new DynamoDBMapper(dynamoDB);
        this.txManager = new TransactionManager(dynamoDB, "Transactions", "TransactionImages");
    }

    public Comment put(Comment comment) {
        Transaction transaction = txManager.newTransaction();

        transaction.save(comment);

        Item item = new Item();
        item.setId(comment.getItemId());

        item = transaction.load(item);
        item.setTotalComments(item.getTotalComments() + 1);
        item.setTotalRating(item.getTotalRating() + comment.getRating());

        transaction.save(item);

        transaction.commit();

        return comment;
    }

    public Comment get(String itemId, String messageId) {
        Comment comment = new Comment();
        comment.setItemId(itemId);
        comment.setMessageId(messageId);

        return mapper.load(comment);
    }

    public void delete(String itemId, String messageId) {
        Comment comment = new Comment();
        comment.setItemId(itemId);
        comment.setMessageId(messageId);

        mapper.delete(comment);
    }

    public List<Comment> getAll() {
        return mapper.scan(Comment.class, new DynamoDBScanExpression());
    }

    public List<Comment> getAllForItem(String itemId) {
        Comment comment = new Comment();
        comment.setItemId(itemId);
        DynamoDBQueryExpression<Comment> queryExpression = new DynamoDBQueryExpression<Comment>()
                .withHashKeyValues(comment);

        return mapper.query(Comment.class, queryExpression);
    }

    public List<Comment> allForItemWithRating(String itemId, int minRating) {
        Comment comment = new Comment();
        comment.setItemId(itemId);
        DynamoDBQueryExpression<Comment> queryExpression = new DynamoDBQueryExpression<Comment>()
                .withHashKeyValues(comment)
                .withRangeKeyCondition(
                        "rating",
                        new Condition()
                                .withComparisonOperator(ComparisonOperator.GE)
                                .withAttributeValueList(new AttributeValue().withN(Integer.toString(minRating)))
                );

        return mapper.query(Comment.class, queryExpression);
    }

    public List<Comment> allForUser(String userId) {
        Comment comment = new Comment();
        comment.setUserId(userId);
        DynamoDBQueryExpression<Comment> queryExpression = new DynamoDBQueryExpression<Comment>()
                .withHashKeyValues(comment)
                .withConsistentRead(false);

        return mapper.query(Comment.class, queryExpression);
    }

}

File: 04/demos/demo/src/main/java/com/pluralsight/dynamodb/dao/CommentDao.kt
package com.pluralsight.dynamodb.dao

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBQueryExpression
import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBScanExpression
import com.amazonaws.services.dynamodbv2.model.AttributeValue
import com.amazonaws.services.dynamodbv2.model.ComparisonOperator
import com.amazonaws.services.dynamodbv2.model.Condition
import com.amazonaws.services.dynamodbv2.transactions.Transaction
import com.amazonaws.services.dynamodbv2.transactions.TransactionManager
import com.pluralsight.dynamodb.domain.Comment
import com.pluralsight.dynamodb.domain.Item

class CommentDao(dynamoDB: AmazonDynamoDB) {

    private val mapper: DynamoDBMapper

    private val txManager: TransactionManager

    val all: List<Comment>
        get() = mapper.scan(Comment::class.java, DynamoDBScanExpression())

    init {
        this.mapper = DynamoDBMapper(dynamoDB)
        this.txManager = TransactionManager(dynamoDB, "Transactions", "TransactionImages")
    }

    fun put(comment: Comment): Comment {
        val transaction = txManager.newTransaction()

        transaction.save(comment)

        var item = Item()
        item.id = comment.itemId

        item = transaction.load(item)
        item.totalComments = item.totalComments + 1
        item.totalRating = item.totalRating + comment.rating

        transaction.save(item)

        transaction.commit()

        return comment
    }

    operator fun get(itemId: String, messageId: String): Comment {
        val comment = Comment()
        comment.itemId = itemId
        comment.messageId = messageId

        return mapper.load(comment)
    }

    fun delete(itemId: String, messageId: String) {
        val comment = Comment()
        comment.itemId = itemId
        comment.messageId = messageId

        mapper.delete(comment)
    }

    fun getAllForItem(itemId: String): List<Comment> {
        val comment = Comment()
        comment.itemId = itemId
        val queryExpression = DynamoDBQueryExpression<Comment>()
                .withHashKeyValues(comment)

        return mapper.query(Comment::class.java, queryExpression)
    }

    fun allForItemWithRating(itemId: String, minRating: Int): List<Comment> {
        val comment = Comment()
        comment.itemId = itemId
        val queryExpression = DynamoDBQueryExpression<Comment>()
                .withHashKeyValues(comment)
                .withRangeKeyCondition(
                        "rating",
                        Condition()
                                .withComparisonOperator(ComparisonOperator.GE)
                                .withAttributeValueList(AttributeValue().withN(Integer.toString(minRating)))
                )

        return mapper.query(Comment::class.java, queryExpression)
    }

    fun allForUser(userId: String): List<Comment> {
        val comment = Comment()
        comment.userId = userId
        val queryExpression = DynamoDBQueryExpression<Comment>()
                .withHashKeyValues(comment)
                .withConsistentRead(false)

        return mapper.query(Comment::class.java, queryExpression)
    }

}
